### DISCUSSION TITLE
needing some more examples of pulling data from a device or interface of a device.

### DISCUSSION THREAD

**Discussion by h3steinhauer** [logicmonitor/lm-powershell-module#6] [Q&A] [answered]
My focus is on the many WAN devices in our LM setup.  We have a central system but we also have some 90+ Facilities that connect to our DataCenter via WAN connections.  

My ask is to be able to pull the WAN Interfaces that are also broken up into VLAN sets as the interface shows 0/0/1.400  for the Vlan 400 link that is shared between all other locations on the same vlan 400 link.   

What I would like to do: 
1 - set those interfaces at the DataCenter to a link speed as determined by the contract.  This is different than the Interface connection speed on the device. 
2 - Set those interfaces at the remote offices to a link speed as determined by the contract. 
3 - be able to pull the usage data at the detail level to study usage over time.  Not summarized by hour. 
4 - we have the typical usage of high during the day with the lunch time lull, but the morning and afternoon peaks need to be studied and watched.    The Overnight usage is very different than the prime time day even though we are a 24x7 operation.  
5 - We need to ensure that our peak load capacity is understood and studied and alerted on for planning. 
6 - it takes time to upgrade and we need lead time to allow for that planning cycle.  

I am thinking I may need to provide an input file that has the device and interfaces that I need to pull data for and that the code would read this file to know which devices to work on.   

Ideally pulling this data weekly for review should be enough to build up the information and study it. But if need be, it could be pulled daily to keep the data as detailed as possible.  

I'm thinking that this may require a secondary database to save off the detail information from the LM gathering process and for further detail study of this usage.   

My question is how to pull the data using the API interface to pull information from selected devices and interfaces and then capture this as CSV information for follow on processing to study. 

How do I construct the API Call to pull the detail data per interface of a device? But also pull it in such a way that I do not overload the API Interface.   The Dashboard Detail has our information at 2 minute intervals.  But when I go to the 7 day view - it summarizes it as 11 minute intervals.  I really need it at the 2 minute interval if possible.  Or I need to pull the data daily to keep it at the 2 minute interval.  

Thanks for your help and guidance on how to pull the data with the Window API Calls?  

Suggestions?

**Comment by stevevillardi** [comment]
You should be able to achieve that fairly easily with the PS module. Here is a sample snippet of code that does most of what you are looking for, you would just need to add the logic to only take action on the vlan specific instances and the code to figure out the contract values for the in and out speed but this script will take a device and pull the instance list for network interfaces and then proceed to set the custom properties for controlling in and out speed along with pulling the last 24 hours of metrics and exporting the entire result set for analysis.

```powershell
#Setup variables
$InterfaceInstanceData = $null
$Results = New-Object System.Collections.ArrayList

#Get device info and datasource/instance list
$Device = Get-LMDevice -Name "192.168.1.4"
$DeviceDS = Get-LMDeviceDatasourceList -id $Device.id -Filter "dataSourceName -eq 'SNMP_Network_Interfaces'"
$DeviceInstances = Get-LMDeviceDatasourceInstance -Deviceid $Device.id -datasourceId $DeviceDS.dataSourceId
Write-Host "Processing Device: $($Device.Name) with $($DeviceInstances.Count) instances"
Foreach($Instance in $DeviceInstances){
    Write-Host "Processing Instance: $($Instance.Name)"
    #Get last 24 hours of instance data
    $InterfaceInstanceData = Get-LMDeviceData -DeviceId $Device.id -DatasourceId $DeviceDS.dataSourceId -InstanceId $Instance.Id

    #Optionally set interface speed based on contract logic would go here
    $UpdatedDeviceInstance = Set-LMDeviceDatasourceInstance -DeviceId $Device.id -DatasourceId $DeviceDS.dataSourceId -InstanceId $Instance.Id -Properties @{out_speed="1000000";in_speed="2000000"}

    #Get updated interface properties, make sure they are set correctly
    $InSpeedIndex = $UpdatedDeviceInstance.customProperties.name.IndexOf("in_speed")
    If($InSpeedIndex -ne -1) {
        $InstanceDownSpeed = $UpdatedDeviceInstance.customProperties.value[$InSpeedIndex]
    }
    $OutSpeedIndex = $UpdatedDeviceInstance.customProperties.name.IndexOf("out_speed")
    If($OutSpeedIndex -ne -1) {
        $InstanceUpSpeed = $UpdatedDeviceInstance.customProperties.value[$OutSpeedIndex]
    }

    #Add Results to Array
    $Results.Add(
        [PSCustomObject]@{
            DeviceName = $Device.Name
            DatasourceName = $DeviceDS.DataSourceName
            InstanceName = $Instance.Name
            InstanceDisplayName = $Instance.displayName
            InstanceDescription = $Instance.description
            InstanceUpSpeed = $InstanceUpSpeed
            InstanceDownSpeed = $InstanceDownSpeed
            Metrics = $InterfaceInstanceData
        }
    ) | Out-Null

}
```

![2024-06-14_14-26-21](https://github.com/logicmonitor/lm-powershell-module/assets/42367049/4eac730f-63d7-427b-bbb5-a3d5588b6823)

**Reply by h3steinhauer** [reply]
I was doing good following you and creating a CSV file of the $results variable. 
But then I hit a snag trying to unpack the Metrics Variable in the PSCustomObject.   That is very cleaver and I see the need for it, but using a simple Export-CSV  did not unpack the overload values in the Metrics Variable. 

My lack of PowerShell experience is very evident.   In other languages, I would pull out that Metrics variable and then unpack it via a series of other statements,  but I'm not sure how to do that in Powershell.     

I think I'm close to being able to do what I have in mind, but I'm missing a key piece of how to unpack that variable - Metrics. 

I notice it that it contains one set of data for each observation that was pulled down.    The general display shows the first observation, but then the rest are hidden in the . . . }  

I'm missing something and it is probably something simple to unpack this.

**Comment by h3steinhauer** [comment, marked as answer]
Thank you for your help - - I learn by example and this is a great example to learn from.   
This shows how to get the data for one Device - and all the interfaces on that device. 

How do I know it is for the last 24 hours?    

I see that output in the variable set   Metrics = $InterfaceInstanceData  

Thank you again for a good example.  This is great stuff and the building blocks for being able to do more with the data that is being collected.

**Reply by stevevillardi** [reply]
By default the Get-LMDeviceData cmdlet will default to last 24 hours if no -StartDate and -EndDate is specified. You specify that parameter with this format:

```powershell
-StartDate $(Get-Date).AddHours(-12) -EndDate $(Get-Date)
```

**Comment by h3steinhauer** [comment]
Another person helped to decode the custom object in this script.  
I'm sharing the function that was created to bring out the variables saved off in the custom object array. 

This is looking for the variable Metrics to then expand. 


# Function to flatten PSCustom Object with nested Metrics 
function Flatten-PSCustomObject {
    param (
        [Parameter(Mandatory = $true)]
        [array]$Objects
    )
 
    $flattenedResults = @()
 
    foreach ($obj in $Objects) {
$commonProperties = $obj.PSObject.Properties | Where-Object { $_.Name -ne 'Metrics' }
        $metrics = $obj.Metrics
 
        foreach ($metric in $metrics) {
            $flattenedObject = @{}
 
            # Add common properties
            foreach ($property in $commonProperties) {
$flattenedObject[$property.Name] = $property.Value
            }
 
            # Add metric properties
foreach ($property in $metric.PSObject.Properties) {
$flattenedObject[$property.Name] = $property.Value
            }
 
            $flattenedResults += [PSCustomObject]$flattenedObject
        }
    }
 
    return $flattenedResults
}

 Flatten-PSCustomObject($results)
